File: java/Wnos_Plugin/src/com/windsor/node/plugin/aqs/DrDasProxyService.java
/*
Copyright (c) 2009, The Environmental Council of the States (ECOS)
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

 * Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.
 * Neither the name of the ECOS nor the names of its contributors may
   be used to endorse or promote products derived from this software
   without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.
 */

/**
 * 
 */
package com.windsor.node.plugin.aqs;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.windsor.node.common.domain.CommonContentType;
import com.windsor.node.common.domain.CommonTransactionStatusCode;
import com.windsor.node.common.domain.DataServiceRequestParameter;
import com.windsor.node.common.domain.Document;
import com.windsor.node.common.domain.NodeTransaction;
import com.windsor.node.common.domain.PaginationIndicator;
import com.windsor.node.common.domain.ProcessContentResult;
import com.windsor.node.common.domain.RequestType;
import com.windsor.node.common.domain.ServiceType;
import com.windsor.node.plugin.BaseWnosPlugin;
import com.windsor.node.service.helper.CompressionService;
import com.windsor.node.service.helper.HeaderDocumentHelper;
import com.windsor.node.service.helper.IdGenerator;
import com.windsor.node.service.helper.RemoteFileResourceHelper;

/**
 * @author mchmarny
 * 
 */
public class DrDasProxyService extends BaseWnosPlugin {

    public static final String SRV_MONITOR = "AQDEMonitorData";
    public static final String SRV_RAW = "AQDERawData";

    public static final String ARG_ENDPOINT_URL = "Reporter Url";
    public static final String ARG_HD_SCHEMA_VERSION = "Schema Version";

    public DrDasProxyService() {

        super();

        debug("Setting internal runtime argument list");

        getConfigurationArguments().put(ARG_ENDPOINT_URL, "");
        getConfigurationArguments().put(ARG_HD_SCHEMA_VERSION, "");

        getSupportedPluginTypes().add(ServiceType.QUERY);
        getSupportedPluginTypes().add(ServiceType.SOLICIT);
        getSupportedPluginTypes().add(ServiceType.QUERY_OR_SOLICIT);

        debug("Plugin initialized");

    }

    /**
     * will be called by the plugin executor after properties are set. an
     * opportunity to validate all settings
     */
    public void afterPropertiesSet() {
        super.afterPropertiesSet();

        debug("Validating runtime args");

        // make sure the run time args are set
        if (getConfigurationArguments() == null) {
            throw new RuntimeException("Data sources not set");
        }

        // make sure the ARG_ENDPOINT_URL arg is set set
        if (!getConfigurationArguments().containsKey(ARG_ENDPOINT_URL)) {
            throw new RuntimeException(ARG_ENDPOINT_URL + " not set");
        }

        debug("Plugin validated");

    }

    /**
     * processlogger.debug
     */
    public ProcessContentResult process(NodeTransaction transaction) {

        debug("Processing transaction...");

        ProcessContentResult result = new ProcessContentResult();
        result.setSuccess(false);
        result.setStatus(CommonTransactionStatusCode.FAILED);

        try {

            result.getAuditEntries()
                    .add(makeEntry("Vaildating transaction..."));

            validateTransaction(transaction);

            IdGenerator idGenerator = (IdGenerator) getServiceFactory()
                    .makeService(IdGenerator.class);

            if (idGenerator == null) {
                throw new RuntimeException("Unable to obtain IdGenerator");
            }

            HeaderDocumentHelper headerHelper = (HeaderDocumentHelper) getServiceFactory()
                    .makeService(HeaderDocumentHelper.class);

            if (headerHelper == null) {
                throw new RuntimeException(
                        "Unable to obtain HeaderDocumentHelper");
            }

            RemoteFileResourceHelper webHelper = (RemoteFileResourceHelper) getServiceFactory()
                    .makeService(RemoteFileResourceHelper.class);

            if (webHelper == null) {
                throw new RuntimeException(
                        "Unable to obtain RemoteFileResourceHelper");
            }

            /*
             * 
             * ARGS
             */
            result.getAuditEntries().add(makeEntry("Acquiring arguments..."));

            String reporterServiceUrl = getRequiredConfigValueAsString(ARG_ENDPOINT_URL);
            String reporterSchemaVersion = getRequiredConfigValueAsString(ARG_HD_SCHEMA_VERSION);

            Map properties = new HashMap();
            properties.put("schemaVersion", reporterSchemaVersion);

            result.getAuditEntries().add(
                    makeEntry("Configuring reporter client using: "
                            + reporterServiceUrl));

            DrDasServiceHelper srv = new DrDasServiceHelper(reporterServiceUrl,
                    transaction.getRequest().getParameters(),
                    reporterSchemaVersion);

            result.getAuditEntries().add(
                    makeEntry("Getting requested service name..."));

            String methodName = transaction.getRequest().getService().getName();

            result.getAuditEntries()
                    .add(
                            makeEntry("Parsing execution strategy from: "
                                    + methodName));

            byte[] resultsBytes = null;

            if (methodName.equalsIgnoreCase(SRV_MONITOR)) {

                result.getAuditEntries().add(
                        makeEntry("Executing Monitor Data service using: "
                                + transaction.getRequest().getParameters()
                                + " and SchemaVersion: "
                                + reporterSchemaVersion));

                resultsBytes = srv.executeMonitorData();

            } else if (methodName.equalsIgnoreCase(SRV_RAW)) {

                result.getAuditEntries().add(
                        makeEntry("Executing Raw Data service using: "
                                + transaction.getRequest().getParameters()
                                + " and SchemaVersion: "
                                + reporterSchemaVersion));

                String resultFile = srv.executeRawData(idGenerator.createId());

                result.getAuditEntries().add(
                        makeEntry("Reading remote result: " + resultFile));

                resultsBytes = webHelper.getBytesFromURL(resultFile);

                String tempXml = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n";
                tempXml += new String(resultsBytes);
                resultsBytes = tempXml.getBytes("UTF-8");

            } else {
                throw new RuntimeException(
                        "Invalid parameter: Method Name is not supported ("
                                + methodName + ").");
            }

            result.getAuditEntries().add(
                    makeEntry("Testing for compression: "
                            + transaction.getRequest().getType()));

            Document doc = new Document();
            result.getAuditEntries().add(makeEntry("Creating document..."));

            if (transaction.getRequest().getType() != RequestType.QUERY) {

                result.getAuditEntries().add(
                        makeEntry("Compressing results..."));

                CompressionService compressionService = (CompressionService) getServiceFactory()
                        .makeService(CompressionService.class);

                doc.setType(CommonContentType.ZIP);
                doc.setDocumentName("AQS-" + methodName + ".zip");
                doc.setContent(compressionService.zip(resultsBytes, "AQS-"
                        + methodName + ".xml"));

            } else {
                doc.setType(CommonContentType.XML);
                doc.setDocumentName("AQS-" + methodName + ".xml");
                doc.setContent(resultsBytes);
            }

            result.getAuditEntries().add(makeEntry("Setting result..."));

            result.setPaginatedContentIndicator(new PaginationIndicator(
                    transaction.getRequest().getPaging().getStart(),
                    transaction.getRequest().getPaging().getCount(), true));

            result.getDocuments().add(doc);

            result.setSuccess(true);
            result.setStatus(CommonTransactionStatusCode.PROCESSED);
            result.getAuditEntries().add(makeEntry("Done: OK"));

        } catch (Throwable ex) {

            error(ex.getMessage(), ex);
            error(ex.getMessage());

            result.setSuccess(false);
            result.setStatus(CommonTransactionStatusCode.FAILED);

            result.getAuditEntries().add(
                    makeEntry("Error while executing: "
                            + this.getClass().getName() + " Message: "
                            + ex.getMessage()));

        }

        return result;
    }

    /*
     * (non-Javadoc)
     * 
     * @see
     * com.windsor.node.plugin.BaseWnosPlugin#getServiceRequestParamSpecs(java
     * .lang.String)
     */
    @Override
    public List<DataServiceRequestParameter> getServiceRequestParamSpecs(
            String serviceName) {

        List<DataServiceRequestParameter> list = new ArrayList<DataServiceRequestParameter>();

        Integer sortIndex = 0;

        if (serviceName.equalsIgnoreCase(SRV_MONITOR)) {

            for (MonitorDataArgType arg : MonitorDataArgType.values()) {

                DataServiceRequestParameter param = new DataServiceRequestParameter();
                param.setName(arg.toString());
                param.setSortIndex(sortIndex);

                list.add(param);

                sortIndex++;
            }

        } else if (serviceName.equalsIgnoreCase(SRV_RAW)) {

            for (RawDataArgType arg : RawDataArgType.values()) {

                DataServiceRequestParameter param = new DataServiceRequestParameter();
                param.setName(arg.toString());
                param.setSortIndex(sortIndex);

                list.add(param);

                sortIndex++;
            }

        } else {

            list = null;
        }

        return list;
    }

}

File: dotnet/prod-opennode2-net/plugins/AQS-WS/DrDasProxyService.cs
#region License
/*
Copyright (c) 2009, The Environmental Council of the States (ECOS)
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

 * Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.
 * Neither the name of the ECOS nor the names of its contributors may
   be used to endorse or promote products derived from this software
   without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.
*/
#endregion

using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using System.Xml;
using System.IO;
using System.Data;
using System.Data.Common;
using System.Data.ProviderBase;
using Windsor.Node2008.WNOSPlugin;
using System.Diagnostics;
using System.Reflection;
using Windsor.Node2008.WNOSUtility;
using Windsor.Node2008.WNOSDomain;
using Windsor.Node2008.WNOSProviders;
using Spring.Data.Common;
using Windsor.Commons.Logging;
using Windsor.Commons.NodeDomain;
using Windsor.Node2008.WNOSPlugin.AQSCommon;

namespace Windsor.Node2008.WNOSPlugin.AQSWS
{

    public enum AQSServiceType
    {
        AQDEMonitorData,
        AQDERawData
    }

    public enum MonitorDataArgType
    {
        FileGenerationPurposeCode,
        SubstanceName,
        MonitorType,
        StateName,
        CountyName,
        CityName,
        TribeName,
        FacilitySiteIdentifier,
        MinLatitudeMeasure,
        MaxLatitudeMeasure,
        MinLongitudeMeasure,
        MaxLongitudeMeasure,
        LastUpdatedDate
    }


    public enum RawDataArgType
    {
        FileGenerationPurposeCode,
        BeginDate,
        BeginTime,
        EndDate,
        EndTime,
        TimeType,
        SampleDuration,
        SubstanceName,
        MonitorType,
        DataValidityCode,
        DataApprovalIndicator,
        StateName,
        CountyName,
        CityName,
        TribeName,
        FacilitySiteIdentifier,
        MinLatitudeMeasure,
        MaxLatitudeMeasure,
        MinLongitudeMeasure,
        MaxLongitudeMeasure,
        LastUpdatedDate,
        IncludeMonitorDetails,
        IncludeEventData
    }




    [Serializable]
    public class DrDasProxyService : AQSBaseHeaderPlugin, ISolicitProcessor, IQueryProcessor
    {

        protected const string CONFIG_REPORTER_URL = "Reporter Url";
        protected const string CONFIG_SCHEMA_VERSION = "Schema Version";

        #region fields
        private static readonly ILogEx LOG = LogManagerEx.GetLogger(MethodBase.GetCurrentMethod());
        private IRequestManager _requestManager;
        private DataRequest _request;

        private string _reporterUrl;
        private string _schemaVersion;
        #endregion

        /// <summary>
        /// Responsible for processing all AQS services
        /// </summary>
        public DrDasProxyService()
        {
            ConfigurationArguments.Add(CONFIG_REPORTER_URL, null);
            ConfigurationArguments.Add(CONFIG_SCHEMA_VERSION, null);
        }

        protected override void LazyInit()
        {
            base.LazyInit();

            GetServiceImplementation(out _requestManager);

            _reporterUrl = ValidateNonEmptyConfigParameter(CONFIG_REPORTER_URL);
            _schemaVersion = ValidateNonEmptyConfigParameter(CONFIG_SCHEMA_VERSION);
        }
        protected virtual void ValidateRequest(string requestId)
        {
            DebugAndAudit("Getting request: " + requestId);
            _request = _requestManager.GetDataRequest(requestId);


            if (_request == null)
            {
                throw new ApplicationException("Null request");
            }

            if (_request.Service == null)
            {
                throw new ApplicationException("Null service");
            }

        }
        #region IQueryProcessor Interface Implementation

        /// <summary>
        /// ProcessQuery
        /// </summary>
        /// <param name="requestId"></param>
        /// <returns></returns>
        public PaginatedContentResult ProcessQuery(string requestId)
        {
            LOG.DebugEnter(MethodBase.GetCurrentMethod(), requestId);

            LazyInit();

            ValidateRequest(requestId);

            string xmlFilePath = GetXmlDocument();

            string zipFilePath = AddExchangeDocumentHeader(xmlFilePath, true, _request.TransactionId);

            DebugAndAudit("Creating paginated content result");
            PaginatedContentResult result =
                new PaginatedContentResult(_request.RowIndex, _request.MaxRowCount, true, CommonContentType.ZIP, 
                                           File.ReadAllBytes(zipFilePath));

            return result;
        }

        #endregion

        #region ISolicitProcessor Interface Implementation

        /// <summary>
        /// ProcessSolicit
        /// </summary>
        /// <param name="requestId"></param>
        public void ProcessSolicit(string requestId)
        {

            LOG.DebugEnter(MethodBase.GetCurrentMethod(), requestId);

            LazyInit();

            ValidateRequest(requestId);

            string xmlFilePath = GetXmlDocument();

            string zipFilePath = AddExchangeDocumentHeader(xmlFilePath, true, _request.TransactionId);
        }

        #endregion

        protected string GetXmlDocument()
        {
            DebugAndAudit("Getting data for: " + _request.Service.Name);
            XmlDocument doc = ProcessServiceRequest(_request);

            string path = _settingsProvider.NewTempFilePath(".xml");

            DebugAndAudit("Loading and saving content");

            using (StringWriter sw = new StringWriter())
            {
                using (XmlTextWriter xw = new XmlTextWriter(sw))
                {
                    doc.WriteTo(xw);
                    UTF8Encoding encoding = new UTF8Encoding(false, true);
                    byte[] content = encoding.GetBytes(sw.ToString());
                    File.WriteAllBytes(path, content);
                }
            }
            return path;
        }
        internal void DebugAndAudit(string message, params object[] args)
        {
            AppendAuditLogEvent(message, args);

            LOG.Debug(message, args);
        }

        /// <summary>
        /// Return the Query, Solicit, or Execute data service parameters for specified data service.
        /// This method should NOT call GetServiceImplementation().
        /// </summary>
        public override IList<TypedParameter> GetDataServiceParameters(string serviceName, out DataServicePublishFlags publishFlags)
        {
            List<TypedParameter> list = null;
            string[] names = null;
            if (string.Equals(serviceName, AQSServiceType.AQDEMonitorData.ToString(), StringComparison.InvariantCultureIgnoreCase))
            {
                names = Enum.GetNames(typeof(MonitorDataArgType));
            }
            else if (string.Equals(serviceName, AQSServiceType.AQDERawData.ToString(), StringComparison.InvariantCultureIgnoreCase))
            {
                names = Enum.GetNames(typeof(RawDataArgType));
            }
            if (names != null)
            {
                list = new List<TypedParameter>(names.Length);
                for (int i = 0; i < names.Length; ++i)
                {
                    list.Add(new TypedParameter(string.Format("P{0} - {1}", (i + 1).ToString("D2"), names[i]),
                                                names[i], true, typeof(string), true));
                }
                publishFlags = DataServicePublishFlags.PublishToEndpointVersion11And20;
            }
            else
            {
                publishFlags = DataServicePublishFlags.DoNotPublish;
            }
            return list;
        }
        /// <summary>
        /// ProcessServiceRequest - returns result path
        /// </summary>
        /// <param name="request"></param>
        /// <returns></returns>
        private XmlDocument ProcessServiceRequest(DataRequest request)
        {

            LOG.DebugEnter(MethodBase.GetCurrentMethod(), request);

            if (request == null || request.Parameters == null)
            {
                throw new ApplicationException("Null request or its parameters");
            }

            DebugAndAudit("Validating configuration arguments...");

            DebugAndAudit("Getting WS Proxy for: {0}", _reporterUrl);

            DebugAndAudit("Parsed Schema Version: {0}", _schemaVersion);

            DebugAndAudit("Initializing DrDAS client...");
            AQDEData client = new AQDEData();
            client.Url = _reporterUrl;
            client.AllowAutoRedirect = true;
            client.SoapVersion = System.Web.Services.Protocols.SoapProtocolVersion.Soap11;
            client.Timeout = 600000;
            client.UseDefaultCredentials = true;

            DebugAndAudit("Parsing target service from: {0}...", request.Service.Name);
            AQSServiceType serviceType = (AQSServiceType)Enum.Parse(typeof(AQSServiceType), request.Service.Name, true);


            XmlDocument doc = new XmlDocument();

            List<string> parameters = GetParameterListSortedByName(request.Parameters);

            switch (serviceType)
            {
                case AQSServiceType.AQDEMonitorData:

                    DebugAndAudit("Executing queryAQSMonitorData...");

                    doc.LoadXml(client.queryAQSMonitorData(
                        parameters[(Int32)MonitorDataArgType.FileGenerationPurposeCode],
                        parameters[(Int32)MonitorDataArgType.SubstanceName],
                        parameters[(Int32)MonitorDataArgType.MonitorType],
                        parameters[(Int32)MonitorDataArgType.StateName],
                        parameters[(Int32)MonitorDataArgType.CountyName],
                        parameters[(Int32)MonitorDataArgType.CityName],
                        parameters[(Int32)MonitorDataArgType.TribeName],
                        parameters[(Int32)MonitorDataArgType.FacilitySiteIdentifier],
                        parameters[(Int32)MonitorDataArgType.MinLatitudeMeasure],
                        parameters[(Int32)MonitorDataArgType.MaxLatitudeMeasure],
                        parameters[(Int32)MonitorDataArgType.MinLongitudeMeasure],
                        parameters[(Int32)MonitorDataArgType.MaxLongitudeMeasure],
                        parameters[(Int32)MonitorDataArgType.LastUpdatedDate],
                        _schemaVersion));

                    break;

                case AQSServiceType.AQDERawData:

                    //Remote request Id
                    string remoteRequestId = Guid.NewGuid().ToString();
                    DebugAndAudit("Remote request Id: {0}", remoteRequestId);
                    DebugAndAudit("Executing solicitAQSRawData...");
                    String resultUrl = client.solicitAQSRawData(
                        parameters[(Int32)RawDataArgType.FileGenerationPurposeCode],
                        parameters[(Int32)RawDataArgType.BeginDate],
                        parameters[(Int32)RawDataArgType.BeginTime],
                        parameters[(Int32)RawDataArgType.EndDate],
                        parameters[(Int32)RawDataArgType.EndTime],
                        parameters[(Int32)RawDataArgType.TimeType],
                        parameters[(Int32)RawDataArgType.SampleDuration],
                        parameters[(Int32)RawDataArgType.SubstanceName],
                        parameters[(Int32)RawDataArgType.MonitorType],
                        parameters[(Int32)RawDataArgType.DataValidityCode],
                        parameters[(Int32)RawDataArgType.DataApprovalIndicator],
                        parameters[(Int32)RawDataArgType.StateName],
                        parameters[(Int32)RawDataArgType.CountyName],
                        parameters[(Int32)RawDataArgType.CityName],
                        parameters[(Int32)RawDataArgType.TribeName],
                        parameters[(Int32)RawDataArgType.FacilitySiteIdentifier],
                        parameters[(Int32)RawDataArgType.MinLatitudeMeasure],
                        parameters[(Int32)RawDataArgType.MaxLatitudeMeasure],
                        parameters[(Int32)RawDataArgType.MinLongitudeMeasure],
                        parameters[(Int32)RawDataArgType.MaxLongitudeMeasure],
                        parameters[(Int32)RawDataArgType.LastUpdatedDate],
                        parameters[(Int32)RawDataArgType.IncludeMonitorDetails],
                        parameters[(Int32)RawDataArgType.IncludeEventData],
                        _schemaVersion,
                        remoteRequestId,
                        Boolean.FalseString);
                    DebugAndAudit("Remote result url: {0}", resultUrl);

                    using (XmlTextReader reader = new XmlTextReader(resultUrl))
                    {
                        reader.Read();
                        doc.Load(reader);
                    }

                    break;

                default:
                    throw new ApplicationException("Invalid service name: " + serviceType);
            }

            return doc;

        }
    }
}
